content/2020-05-22-book-review-writing-an-interpreter-in-go/index.md

```diff
@@ -0,0 +1,581 @@
+>
+> -Ball 132
+
+I found this super helpful (at least comparing it to say Rust). Interpreter has one-step. Compilers have two.
+
+#### tree-walking interpreters
+
+I wrote down this because I wanted to look up a few examples of programming languages that fall into this category. I don't know how trustworthy [these Reddit comments](https://www.reddit.com/r/ProgrammingLanguages/comments/dfi9lk/viability_of_treewalk_interpreters/) are but here are a few notes:
+
+> Ruby was a tree-walk interpreter before version 1.9 (I think) and was used in production quite a bit, albeit with a reputation for being very slow and consuming a lot of RAM.
+
+> R was also a tree-walking interpreter for around 20 years! It only changed in the last 5 or so years, a similar time as Ruby.
+
+> It seems Perl still is a tree-walk interpreter.
+
+A bit later, Thorsten does touch on a few. He mentions the Ruby example.
+
+#### intermediate representation
+
+I hadn't heard this term before. According to [Wikipedia](https://en.wikipedia.org/wiki/Intermediate_representation),
+
+> intermediate representation is the [data structure](https://en.wikipedia.org/wiki/Data_structure) or code used internally by a [compiler](https://en.wikipedia.org/wiki/Compiler) or [virtual machine](https://en.wikipedia.org/wiki/Virtual_machine) to represent [source code](https://en.wikipedia.org/wiki/Source_code).
+
+#### just in time
+
+He also mentions on page 133 a JIT, or "just in time" interpreter/compiler. From what I can tell, [JavaScript](https://blog.sessionstack.com/how-javascript-works-inside-the-v8-engine-5-tips-on-how-to-write-optimized-code-ac089e62b12e) usually uses this technique.
+
+One thing to note as well,
+
+> An interpreter that compiles to byte code and uses a virtual machine to evaluate said bytecode is going to be a lot faster.
+>
+> -Ball 133
+
+This is compared to a "tree-walking interpreter that recursively evaluates an AST."
+
+He also mentions WebKit JavaScript engine JavaScriptCore and the interpreter "SquirrelFish" (weird name, right?). This one has "four different stages of JIT compilation."
+
+Lua is another languages that adopted a JIT, but 12 years after the first release (Ball 134).
+
+#### host language
+
+It was interesting that for our evaluator, Thorsten said we only needed two things,
+
+> a tree-walking evaluator and a way to represent Monkey values in our host language Go.
+>
+> -Ball 135
+
+This stuck out to me because I realized Go, the language in which we were writing the language, is this said host language. It made me think of Reason, whose host language is OCaml. But I believe when you look at Rust, it's host language is Rust. And same with TypeScript whose host language is TypeScript. Weird, but cool!
```

Comment:
That's why they're called self-hosted compilers 🙂

Small typo:

```suggestion
This stuck out to me because I realized Go, the language in which we were writing the language, is this said host language. It made me think of Reason, whose host language is OCaml. But I believe when you look at Rust, its host language is Rust. And same with TypeScript whose host language is TypeScript. Weird, but cool!
```

Reply:
🤯 I had no idea! Adding that in here as a side note